diff --git a/src/duk_config.h b/src/duk_config.h
index a4e997e..e12c46e 100644
--- a/src/duk_config.h
+++ b/src/duk_config.h
@@ -336,12 +336,20 @@
 #define DUK_F_CPP11
 #endif
 
+/* NeXT / OPENSTEP */
+#if defined(NeXT) || defined(__NeXT__) || defined(__NeXT)
+#define DUK_F_NO_STDINT_H
+#define DUK_F_OPENSTEP
+#endif
+
 /* GCC.  Clang also defines __GNUC__ so don't detect GCC if using Clang. */
 #if defined(__GNUC__) && !defined(__clang__) && !defined(DUK_F_CLANG)
 #define DUK_F_GCC
 #if defined(__GNUC__) && defined(__GNUC_MINOR__) && defined(__GNUC_PATCHLEVEL__)
 /* Convenience, e.g. gcc 4.5.1 == 40501; http://stackoverflow.com/questions/6031819/emulating-gccs-builtin-unreachable */
 #define DUK_F_GCC_VERSION  (__GNUC__ * 10000L + __GNUC_MINOR__ * 100L + __GNUC_PATCHLEVEL__)
+#elif defined(__GNUC__) && defined(__GNUC_MINOR__)
+#define DUK_F_GCC_VERSION  (__GNUC__ * 10000L + __GNUC_MINOR__ * 100L)
 #else
 #error cannot figure out gcc version
 #endif
@@ -867,6 +875,14 @@
 #define DUK_USE_OS_STRING "unknown"
 #endif  /* autodetect platform */
 
+#if defined(NeXT) || defined(__NeXT__) || defined(__NeXT)
+#undef DUK_USE_DATE_TZO_GMTIME_R
+#define DUK_USE_DATE_TZO_GMTIME
+#undef DUK_USE_DATE_PRS_STRPTIME
+#define DUK_USE_COMPUTED_INFINITY
+#define DUK_DOUBLE_INFINITY  duk_computed_infinity
+#endif
+
 /* Shared includes: C89 */
 #include <stdio.h>
 #include <stdlib.h>
@@ -877,6 +893,33 @@
 #include <math.h>
 #include <limits.h>
 
+#if defined(DUK_F_OPENSTEP)
+static int duk__openstep_vsnprintf(char *str, size_t size, const char *format, va_list ap) {
+	char tmpbuf[65536];
+	int rc;
+
+	rc = vsprintf(tmpbuf, format, ap);
+	if (size > 0) {
+		strncpy(str, tmpbuf, size - 1);
+		str[size - 1] = (char) 0;
+	}
+	return rc;
+}
+
+static int duk__openstep_snprintf(char *str, size_t size, const char *format, ...) {
+	va_list ap;
+	int rc;
+
+	va_start(ap, format);
+	rc = duk__openstep_vsnprintf(str, size, format, ap);
+	va_end(ap);
+	return rc;
+}
+
+#define snprintf duk__openstep_snprintf
+#define vsnprintf duk__openstep_vsnprintf
+#endif
+
 /* date.h is omitted, and included per platform */
 
 /* Shared includes: stdint.h is C99 */
@@ -1147,7 +1190,9 @@
 #endif
 #elif defined(DUK_F_GCC)
 /* --- GCC --- */
-#if defined(DUK_F_C99) || defined(DUK_F_CPP11)
+#if defined(NeXT) || defined(__NeXT__) || defined(__NeXT)
+#define DUK_VA_COPY(dest,src) do { (dest) = (src); } while (0)
+#elif defined(DUK_F_C99) || defined(DUK_F_CPP11)
 /* C99 / C++11 and above: rely on va_copy() which is required. */
 #define DUK_VA_COPY(dest,src) va_copy(dest,src)
 #else
